packages/cactus-plugin-ledger-connector-ethereum/src/main/typescript/plugin-ledger-connector-ethereum.ts

```diff
@@ -953,24 +961,25 @@ export class PluginLedgerConnectorEthereum
     if (isWeb3SigningCredentialNone(req.web3SigningCredential)) {
       throw new Error(`Cannot deploy contract with pre-signed TX`);
     }
+    return this.deployContractFromKeychain(req, req.contract);
+  }
 
-    if (isContractJsonDefinition(req.contract)) {
-      return this.runContractDeployment({
-        web3SigningCredential: req.web3SigningCredential,
-        gasConfig: req.gasConfig,
-        constructorArgs: req.constructorArgs,
-        value: req.value,
-        contractJSON: req.contract.contractJSON,
-      });
-    } else if (isContractKeychainDefinition(req.contract)) {
-      return this.deployContractFromKeychain(req, req.contract);
-    } else {
-      // Exhaustive check
-      const unknownContract: never = req.contract;
-      throw new Error(
-        `Unknown contract definition provided: ${unknownContract}`,
-      );
+  public async deployContractNoKeychain(
```

Comment:
I think I'd prefer to keep a single deploy endpoint for both scenarios with / without a keychain.
- This is roughly the same for the user (no keychain - don't use keychain option in endpoints, if you do the error will be thrown on runtime. This is similar to error thrown when you use regular deploy instead of deployContractNoKeychain)
- This will spread to other methods that may use keychain in the future, and will require both method flawors in order to work (invokeNoKeychain etc..). Also similar strategy must be assumed in the future, if we support other optional plugins in methods, like DB.
- We could use default in-memory keychain plugin in case user doesn't provide one, this would simplify implementations in the future but I'm not sure if this is valid default security-wise.
- These comments are mostly for ethereum connector.

@petermetz You are the creator of this issue, any thoughts on it?

Reply:
@outSH TLDR: I recommend re-using code in the form of smaller functions under the hood but keeping the endpoints separate so that the validation logic is straight forward. The guiding principle being that we want to reduce if conditions because they are complexity. (I was talking to @jagpreetsinghsasan about the same thing the other day).

The obvious downside is the strong potential for code repetition and some of it will be unavoidable but I believe that with the right structuring of the code in smaller functions we can avoid 95% of that and the remaining 5% is just worth it as a trade-off.

Of course if we find out while writing the actual code that the 5/95 ratio is more like 50/50 then we should reevaluate our choices but I'm hopeful that we can collaborate and get to the (currently just imagined) 5/95 ratio.

## Code Re-use Example

This is what I mean roughly speaking as a vision (it is very similar to my other message on the connector inheritance thread):

```typescript
class DeployEndpointX {
   handleRequest() {
    // ...
   }
}


class DeployEndpointY {
   handleRequest() {
    // ...
   }
}

class Connector {

   deployX() {
      this.smallerReusableFunction1();
      this.smallerReusableFunction2();
   }

   // Mostly the same as deployX() but
   // it has some more logic.
   // No need for if conditions though
   // which is nice
   deployY() {
      this.smallerReusableFunction1();
      this.smallerReusableFunction2();
      this.smallerReusableFunction3();
   }

   smallerReusableFunction1() {

   }

   smallerReusableFunction2() {
      
   }

   smallerReusableFunction3() {
      
   }

   smallerReusableFunction4() {
      
   }
}
```


## Specifically on validation

The other big benefit of separate endpoints is that the validation can be locked down much tighter.
If our request definitions in the OpenAPI specs are designed to handle multiple different use-cases (such as keychain & no-keychain) then we are either
1. Forced to mark some of the properties non-mandatory just because we don't know at design time whether it's mandatory for that code path or not (results in more if conditions within the implementation just to check this as well)
2. Use inheritance (and still handle that with switch cases or if conditions too) which is poorly supported by a lot of the code templates that the generator has (we currently have probably 80% of the code generation related issues due to class inheritance not being fully supported either by the generator itself or by the template that we'd like to use)

Reply:
@petermetz

For me, endpoints are supposed to define single business action without technical details on how it's implemented. I'm still afraid this will burst exponentially into more specialty methods like that if we add more features (another auth method, another storage for contracts, etc..). As for openapi, right now multiple options can be supported with `oneOf` which works in TS and is informative for the user, I don't think we should mess our API just because the tooling requires us to do so.

Reply:
> @petermetz
> 
> For me, endpoints are supposed to define single business action without technical details on how it's implemented. I'm still afraid this will burst exponentially into more specialty methods like that if we add more features (another auth method, another storage for contracts, etc..). As for openapi, right now multiple options can be supported with `oneOf` which works in TS and is informative for the user, I don't think we should mess our API just because the tooling requires us to do so.

@outSH Sorry for the very slow response. I had think long and deep about this one. Based on the comments so far I think we are both pretty set on our viewpoints which happen to be the exact opposite of each other. How about the following plan:
1. You'll be the principal maintainer of `cactus-plugin-ledger-connector-ethereum` and consequently we leave that out of the keychain dependency refactor because that's how you prefer it.
2. I'll be the principal maintainer of `cactus-plugin-ledger-connector-besu` and we'll make the keychain optional there because that's just how I'd prefer it (not exactly with the current diff of this PR because I see some things I would change still but me and Jagpreet we'll push through that together)
3. The xdai and quorum seem pretty dead so TBH we can just remove them IMO and then it's that much less code to maintain / have this conversation about.

Does this seem like a reasonable compromise?

Reply:
@petermetz I'd prefer to have these connectors merged some day (i.e. use common base code in the future, as described in connector inheritance discussion some time ago) so I'd prefer to stay in line as much as possible. If you're sure we want to go with separate endpoinds then we can do that :) We'll just have to keep an eye on how this progresses in the future as you've mentioned before.

Reply:
Oh, and in general +1 to make this PR smaller and easier to review and reason about :)

Reply:
> @petermetz I'd prefer to have these connectors merged some day (i.e. use common base code in the future, as described in connector inheritance discussion some time ago) so I'd prefer to stay in line as much as possible. If you're sure we want to go with separate endpoinds then we can do that :) We'll just have to keep an eye on how this progresses in the future as you've mentioned before.

@outSH Got it, thank you for the compromise! Alright, that's our plan then. We can plan on doing the merging together for Cacti v3.0.0 (because we'll be able to do breaking changes that time again) and apply all that we've learned from the current refactors (could be I'll become a convert and start advocating for your endpoint design too, it happens often that once I write the code for something my opinion changes).